android/src/com/mapswithme/maps/location/TrackRecorder.java

```diff
@@ -0,0 +1,217 @@
+{
+  INSTANCE;
+
+  @NonNull
+  private static final Logger LOGGER = LoggerFactory.INSTANCE.getLogger(LoggerFactory.Type.TRACK_RECORDER);
+
+  private static final String TAG = TrackRecorder.class.getSimpleName();
+
+  private static final long WAKEUP_INTERVAL_MS = 20000;
+  private static final long STARTUP_AWAIT_INTERVAL_MS = 5000;
+
+  private static final String LOCATION_TIMEOUT_STORED_KEY = "TrackRecordLastAwaitTimeout";
+  private static final long LOCATION_TIMEOUT_MIN_MS = 5000;
+  private static final long LOCATION_TIMEOUT_MAX_MS = 80000;
+
+  private boolean mInitialized = false;
+
+  @SuppressWarnings("NotNullFieldNotInitialized")
+  @NonNull
+  private Context mContext;
+  @NonNull
+  private AlarmManager mAlarmManager;
+  @NonNull
+  private final Runnable mStartupAwaitProc = this::restartAlarmIfEnabled;
+  @NonNull
+  private final LocationListener mLocationListener = new LocationListener.Simple()
+  {
+    @Override
+    public void onLocationUpdated(Location location)
+    {
+      LOGGER.d(TAG, "onLocationUpdated()");
+      setAwaitTimeout(LOCATION_TIMEOUT_MIN_MS);
+      TrackRecorderWakeService.stop();
+    }
+
+    @Override
+    public void onLocationError(int errorCode)
+    {
+      LOGGER.e(TAG, "onLocationError() errorCode: " + errorCode);
+      // Unrecoverable error occured: GPS disabled or inaccessible
+      setEnabled(false);
+    }
+  };
+
+  @Override
+  public void initialize(@Nullable Context context)
+  {
+    LOGGER.d(TAG, "Initialization of track recorder and setting the listener for track changes");
+    mContext = context;
+    mAlarmManager = (AlarmManager) MwmApplication.from(context)
+                                                 .getSystemService(Context.ALARM_SERVICE);
+
+    MwmApplication.backgroundTracker(context).addListener(foreground -> {
+      LOGGER.d(TAG, "Transit to foreground: " + foreground);
+
+      UiThread.cancelDelayedTasks(mStartupAwaitProc);
+      if (foreground)
+        TrackRecorderWakeService.stop();
+      else
+        restartAlarmIfEnabled();
+    });
+
+    if (nativeIsEnabled())
+      UiThread.runLater(mStartupAwaitProc, STARTUP_AWAIT_INTERVAL_MS);
+    else
+      stop();
+
+    mInitialized = true;
+  }
+
+  @Override
+  public void destroy()
+  {
+    // No op.
+  }
+
+  private void checkInitialization()
+  {
+    if (!mInitialized)
+      throw new AssertionError("Track recorder is not initialized!");
+  }
+
+  private PendingIntent getAlarmIntent()
+  {
+    Intent intent = new Intent(MwmApplication.from(mContext), TrackRecorderWakeReceiver.class);
+    return PendingIntent.getBroadcast(MwmApplication.from(mContext), 0, intent, PendingIntent.FLAG_IMMUTABLE);
+  }
+
+  private void restartAlarmIfEnabled()
+  {
+    LOGGER.d(TAG, "restartAlarmIfEnabled()");
+    if (nativeIsEnabled())
+      mAlarmManager.set(AlarmManager.ELAPSED_REALTIME_WAKEUP, SystemClock.elapsedRealtime() + WAKEUP_INTERVAL_MS, getAlarmIntent());
```

Comment:
Постоянный запуск аларма для записи выглядит адским костылём, несовместимым с батареей.
Читал вот этот [тред](https://github.com/osmandapp/Osmand/issues/5632)? 

Давай записывать для начала в пассивном режиме, как обсуждали, только от других апп, от системы и свои активные позиции. Пускай будет от системы раз-два в час, для записи следа достаточно.

> If background location access is essential for your app, keep in mind that Android preserves device battery life by setting background location limits on devices that run Android 8.0 (API level 26) and higher. On these versions of Android, if your app is running in the background, it can receive location updates only a few times each hour.

Reply:
Ну вообщем тогда и называть вещи своими именами. Не восстановить track recorder, а написать заново.

Reply:
Сколько он сейчас жрёт батареи? Проверь, пожалуйста, на нескольких устройствах.

Reply:
Сейчас он всё равно почему-то в фоне нормально не пишет :(  Ну что же, будем дальше копать.

Reply:
Видимо потому, что лазейку с Alarm прикрыли, она слишком очевидная и жрущая батарею. Давай сделаем по "пассивной" схеме, как я описал. Тем, кто часто смотрит в карту, будет юзабельно.